rust/matches/src/lib.rs

```diff
@@ -623,6 +626,146 @@ pub mod matches {
 
         Ok(())
     }
+
+    pub fn match_join_namespace<'a, 'b, 'c, 'info>(
+        ctx: Context<'a, 'b, 'c, 'info, MatchJoinNamespace<'info>>,
+    ) -> Result<()> {
+        if !is_namespace_program_caller(&ctx.accounts.instructions.to_account_info()) {
+            return Err(error!(ErrorCode::UnauthorizedCaller));
+        }
+
+        let match_instance = &mut ctx.accounts.match_instance;
+
+        let namespaces = match match_instance.namespaces.clone() {
+            Some(namespaces) => namespaces,
+            None => return Err(error!(ErrorCode::FailedToJoinNamespace)),
+        };
+
+        let mut joined = false;
+        let mut new_namespaces = vec![];
+        for mut ns in namespaces {
+            if ns.namespace == anchor_lang::solana_program::system_program::id() && !joined {
+                ns.namespace = ctx.accounts.namespace.key();
+                ns.index = None;
+                ns.inherited = InheritanceState::NotInherited;
+                joined = true;
+                new_namespaces.push(ns);
+            } else {
+                new_namespaces.push(ns);
+            }
+        }
+        if !joined {
+            return Err(error!(ErrorCode::FailedToJoinNamespace));
+        }
+        match_instance.namespaces = Some(new_namespaces);
+
+        Ok(())
+    }
+
+    pub fn match_leave_namespace<'a, 'b, 'c, 'info>(
+        ctx: Context<'a, 'b, 'c, 'info, MatchLeaveNamespace<'info>>,
+    ) -> Result<()> {
+        if !is_namespace_program_caller(&ctx.accounts.instructions.to_account_info()) {
+            return Err(error!(ErrorCode::UnauthorizedCaller));
+        }
+
+        let match_instance = &mut ctx.accounts.match_instance;
+
+        let namespaces = match match_instance.namespaces.clone() {
+            Some(namespaces) => namespaces,
+            None => return Err(error!(ErrorCode::FailedToLeaveNamespace)),
+        };
+
+        let mut left = false;
+        let mut new_namespaces = vec![];
+        for mut ns in namespaces {
+            if ns.namespace == ctx.accounts.namespace.key() && !left {
+                // if the artifact is still cached, error
+                if ns.index != None {
+                    return Err(error!(ErrorCode::FailedToLeaveNamespace));
+                };
+                ns.namespace = anchor_lang::solana_program::system_program::id();
+                ns.inherited = InheritanceState::NotInherited;
+                left = true;
+                new_namespaces.push(ns);
+            } else {
```

Comment:
Same as above, remove the else

Reply:
It's either this or use a continue, imo i think the else is more clear